test/functional/blockprod_generate_blocks_all_sources.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+"""Mempool reorg test
+
+Check that:
+* Transactions are collected into blocks when a new block is issued.
+* Transactions are correctly put back into mempool when the block is reorged out.
+"""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.mintlayer import *
+
+import time
+
+class GenerateBlocksFromAllSourcesTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        self.extra_args = [[
+            "--blockprod-min-peers-to-produce-blocks=0",
+        ]]
+
+    def run_test(self):
+        node = self.nodes[0]
+
+        block_input_data = block_input_data_obj.encode(
+            {
+                "PoW": {
+                    "reward_destination": "AnyoneCanSpend",
+                }
+            }
+        ).to_hex()[2:]
+
+        (generate_utxos, utxo_id) = make_tx(
+            [reward_input(node.chainstate_best_block_id())],
+            [1000 for _ in range(12)],
+        )
+
+        block = node.blockprod_generate_block(block_input_data, [generate_utxos], [], False)
+        node.chainstate_submit_block(block)
+        utxos = [(utxo_id, i) for i in range(12)]
+
+        for include_transaction in ([True, False]):
+            for include_transaction_id in ([True, False]):
+                for include_mempool in ([True, False]):
```

Comment:
I like covering all cases, but I see below only the preparation, not testing the outcome. I would expect the test to first prepare the case (as you're doing below, maybe more if necessary), but also then to test the final result and ensure it's compatible with the expectation. For example, if we're not supposed to take transactions from the mempool, there should be a transaction in the mempool and we should prove that the code didn't pull that transaction from the mempool. Then when the mempool flag is enabled, it would pull that transaction from the mempool and we should see it as part of the final block that was produced.

So before calling `node.chainstate_submit_block(block)`, we should check the contents of the block and ensure that the combination of `include_transaction`, `include_transaction_id` and `include_mempool` was obeyed. Even if that means creating random transactions on every iteration. 

Note that the only issue with this (perfect) plan is that there's no functionality now to evict the mempool to recreate a clean situation on every iteration. A trick around that is to restart the node; as it'll lose all transactions since there's no network. The functionality to restart the node is available in functional tests, but I'm not sure whether it's complete (it's available from bitcoin, but not sure how adapted it's to our code). Worse-case scenario, write a TODO on it + create an issue until the eviction from the mempool is available.

Reply:
Yep, makes sense. I'll see if I can find the restart and get it working here.

Reply:
`stop_node()` and `start_node()` did the trick to flush the mempool.